### DISCUSSION TITLE
Why is `{ load: LoadOp::Load, store: StoreOp::Discard }` treated as `read_only: false`?

### DISCUSSION THREAD

**Discussion by LukasKalbertodt** [gfx-rs/wgpu#4956] [Q&A]
Setting the following in my `RenderPassDescriptor::depth_stencil_attachment`:

```rust
                depth_ops: Some(wgpu::Operations {
                    load: wgpu::LoadOp::Load,
                    store: wgpu::StoreOp::Discard,
                }),
```

Results in an error because I also want to read from the same texture in my shader. I dug a bit into the code and I think it comes down to this:

https://github.com/gfx-rs/wgpu/blob/e26a85365484411f42bc5ddbc08c3a069bba2f83/wgpu/src/backend/direct.rs#L419-L448

It makes sense to me that `LoadOp::Clear` as well as `StoreOp::Store` would both write to the texture. But the above `Operations` seems `read_only: true` to me, so why is it treated as `read_only: false`?

---

I can make this work simply by setting `depth_ops: None`. But I only came up with this after reading the above code. So I'm curious.

**Comment by cwfitzgerald** [comment]
It's perfectly valid to do Load -> Discard and still write to the attachment. An example could be a depth buffer that you used in a previous pass, will do rendering against, but then don't need again.

**Reply by LukasKalbertodt** [reply]
I'm not sure I understand :/

So in one render pass, I add this specific texture as depth attachment with `(Load, Discard)`. The pipeline performs z-tests against it, but `DepthStencilState::depth_write_enabled` is `false`.  I also have it in a bind group in order to read it from shader. So nothing in this render pass changes the texture and this should be fine, right? But it causes the `ResourceUsageConflict` error.
Am I misunderstanding `LoadOp::Load`? It just means "don't clear the texture before using it", right? What even is the difference between `None` and `Some(wgpu::Operations { load: LoadOp::Load, store: StoreOp::Discard })`?

Sorry for taking your time and thanks for the help.

**Reply by cwfitzgerald** [reply]
> Sorry for taking your time and thanks for the help.

Don't worry about it!

I think the important part to understand is that we need to make the read-only/read-write decision _at the beginning_ of the renderpass, based only on the operations provided. While in theory we could try to determine if the attachment is written to in the body of the renderpass, that would require us to do a full pass over the renderpass before we start validation of it - in general WebGPU prefers asking the user for the information and then validating it inline.

> What even is the difference between None and Some(wgpu::Operations { load: LoadOp::Load, store: StoreOp::Discard })?

The difference is if you're _allowed_ to use read/write pipelines. If None, you're not allowed, if Some you are.

**Reply by LukasKalbertodt** [reply]
Hu, interesting.

```rust
let pipeline = gfx.device.create_render_pipeline(&wgpu::RenderPipelineDescriptor {
    depth_stencil: Some(wgpu::DepthStencilState {
        depth_write_enabled: true,
        ...
    }),
    ...
});

let mut rp = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
    depth_stencil_attachment: Some(wgpu::RenderPassDepthStencilAttachment {
        depth_ops: Some(wgpu::Operations {
            load: wgpu::LoadOp::Load,
            store: wgpu::StoreOp::Discard,
        }),
        ...
    }),
    ...
});
```

That code does not result in a validation error. And that's intended? I guess it's useful to re-use pipelines that usually write depth, but in a render pass you can override that decision and say "not this time" basically? Then maybe I understood it.
